### PR TITLE
Introduce Cluster Connectivity for the gRPC Client

### PR DESCRIPTION
**Pull request by pgermishuys**
Repurposed the [endpoint discoverer](https://github.com/EventStore/EventStore/blob/master/src/EventStore.ClientAPI/Internal/ClusterDnsEndPointDiscoverer.cs) from the TCP Client API for node discovery.

Here is an example configuration for connecting to a cluster of nodes with a provided set of gossip seeds.
```
var client = new EventStoreClient(new EventStoreClientSettings
{
    ConnectivitySettings =
    {
        GossipSeeds = new[]
        {
            new DnsEndpoint("localhost", 1114),
            new DnsEndpoint("localhost", 2114),
            new DnsEndpoint("localhost", 3114),
        },
        NodePreference = NodePreference.Leader,
        MaxDiscoverAttempts = 5,
    }
};
```

### CODE REVIEW

File: src/EventStore.Client.Common/Constants.cs [hunk 1]
```diff
@@ -46,6 +46,7 @@ public static class Headers {
 			public const string BasicScheme = "Basic";
 
 			public const string ConnectionName = "connection-name";
+			public const string RequiresMaster = "requires-master";
```
Comment (thefringeninja) [inline src/EventStore.Client.Common/Constants.cs]: This is the old terminology

File: src/EventStore.Client.Tests/NodeBasedClientHandlerTests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using EventStore.Client.EventStore.Client.ConnectionStrategies;
+using Xunit;
+
+namespace EventStore.Client {
+	public class NodeBasedClientHandlerTests {
+		[Theory, InlineData(true), InlineData(false)]
+		public async Task should_set_requires_master_header(bool requiresMaster) {
```
Comment (thefringeninja) [inline src/EventStore.Client.Tests/NodeBasedClientHandlerTests.cs]: Legacy terminology

File: src/EventStore.Client/EventStoreClientConnectivitySettings.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+
+namespace EventStore.Client {
+	public class EventStoreClientConnectivitySettings {
+		public int MaxDiscoverAttempts { get; set; }
+		public Uri[] GossipSeeds { get; set; } = Array.Empty<Uri>();
+		public TimeSpan GossipTimeout { get; set; }
+		public TimeSpan DiscoveryInterval { get; set; }
+		public NodePreference NodePreference { get; set; }
+		public bool RequiresMaster { get; set; }
```
Comment (thefringeninja) [inline src/EventStore.Client/EventStoreClientConnectivitySettings.cs]: Legacy terminology

File: src/EventStore.Client/NodeBasedClientHandler.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+using EventStore.Client.EventStore.Client.ConnectionStrategies;
+
+namespace EventStore.Client {
+	public class NodeBasedClientHandler : DelegatingHandler {
+		private static SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
+		private readonly bool _requiresMaster;
+		private readonly IEndpointDiscoverer _endpointDiscoverer;
+		private NodeEndpoint _endpoint;
+
+		public NodeBasedClientHandler(bool requiresMaster, IEndpointDiscoverer endpointDiscoverer) {
```
Comment (thefringeninja) [inline src/EventStore.Client/NodeBasedClientHandler.cs]: Legacy terminology

File: src/EventStore.Core/Services/Transport/Grpc/ServiceBase.cs [hunk 5]
```diff
@@ -72,6 +72,12 @@ public class ServiceBase {
 				return true;
 			}
 		}
+		
+		public static bool GetRequiresMaster(Metadata requestHeaders) {
```
Comment (thefringeninja) [inline src/EventStore.Core/Services/Transport/Grpc/ServiceBase.cs]: Legacy terminology

File: src/EventStore.Client/GossipBasedEndpointDiscoverer.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		}
+
+		private NodeEndpoint TryDetermineBestNode(IEnumerable<ClusterMessages.MemberInfo> members,
+			NodePreference nodePreference) {
+			var notAllowedStates = new[] {
+				ClusterMessages.VNodeState.Manager,
+				ClusterMessages.VNodeState.ShuttingDown,
+				ClusterMessages.VNodeState.Shutdown,
+				ClusterMessages.VNodeState.Manager,
+				ClusterMessages.VNodeState.Shutdown,
```
Comment (ahjohannessen) [inline src/EventStore.Client/GossipBasedEndpointDiscoverer.cs]: dublicate `ShutDown`

File: src/EventStore.Client/GossipBasedEndpointDiscoverer.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+			if (node == default(ClusterMessages.MemberInfo)) {
+				return null;
+			}
+
+			var nodeEndpoint = new UriBuilder(Uri.UriSchemeHttps, node.ExternalHttpIp, node.ExternalHttpPort).Uri;
+			return new NodeEndpoint(node.InstanceId, nodeEndpoint);
+		}
+
+		private bool IsReadOnlyReplicaState(ClusterMessages.VNodeState state) {
+			return state == ClusterMessages.VNodeState.ReadOnlyLeaderless
```
Comment (ahjohannessen) [inline src/EventStore.Client/GossipBasedEndpointDiscoverer.cs]: `notAllowedStates` contains `ReadOnlyLeaderless` and it is filtered out in `nodes`

**Comment by ahjohannessen** [comment]
Why are membernodes not exposed via grpc endpoint streams instead of json and http pulling?

**Comment by ahjohannessen** [comment]
Just some thoughts from the top of my head:

```protobuf
service ClusterService {
  rpc endpoints(EndpointRequest) returns (stream GlusterInfo);
}

message ClusterInfo {
  MemberInfo          best_node = 1;
  repeated MemberInfo members   = 2;
}

message EndpointRequest {

 enum NodePreference {
  Leader = 0;
  ...
 }

  NodePreference preference = 1;
  ...
}

message MemberInfo {

  enum VNodeState {
    Initializing       = 0;
    Unknown            = 1;
    PreReplica         = 2;
    CatchingUp         = 3;
    Clone              = 4;
    Follower           = 5;
    PreLeader          = 6;
    Leader             = 7;
    Manager            = 8;
    ShuttingDown       = 9;
    Shutdown           = 10;
    ReadOnlyLeaderless = 11;
    PreReadOnlyReplica = 12;
    ReadOnlyReplica    = 13;
    ResigningLeader    = 14;
  }

  UUID       instance_id        = 1;
  VNodeState state              = 2;
  bool       is_alive           = 3;
  string     external_http_ip   = 4;
  uint32     external_http_port = 5;
}
```

File: src/EventStore.Core/Services/Transport/Grpc/ServiceBase.cs [hunk 8]
```diff
@@ -73,6 +75,12 @@ public class ServiceBase {
 			}
 		}
 
+		public static bool GetRequiresLeader(Metadata requestHeaders) {
+			var requiresLeader =
+				requestHeaders.FirstOrDefault(x => x.Key == Constants.Headers.RequiresLeader)?.Value;
+			return !string.IsNullOrEmpty(requiresLeader) && bool.Parse(requiresLeader);
```
Comment (thefringeninja) [inline src/EventStore.Core/Services/Transport/Grpc/ServiceBase.cs]: Maybe use TryParse Here instead

File: src/EventStore.Client/ClusterAwareHttpHandler.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+				Interlocked.CompareExchange(ref _endpoint,
+					new Lazy<Task<IPEndPoint>>(() => _endpointDiscoverer.DiscoverAsync(),
+						LazyThreadSafetyMode.ExecutionAndPublication), endpointResolver);
+
+				throw;
+			}
+		}
+
+		public void ExceptionOccurred(Exception exception) {
+			if (exception is NotLeaderException) {
```
Comment (thefringeninja) [inline src/EventStore.Client/ClusterAwareHttpHandler.cs]: `exception is NotLeaderException lex`

**Pull request merged**